### PR TITLE
[Admin][ProductOptions] Changed form ProductOptionChoiceType to Produ…

### LINKED ISSUE #13742: Reaching memory limit when creating configurable product.
**Sylius version affected**: 1.10.7

**Description**  
Reaching memory limit when creating configurable product.
When creating new configurable product, sylius tries to load all existing product options at once and by doing that you can reach your memory limit.

**Steps to reproduce**  
Create multiple options in you DB table `sylius_product_option`
Visit `sylius_admin_product_create` route
If you created enough product options or you have low memory limit, you should get an error `Fatal error: Allowed memory size of *NUMBER* bytes exhausted `

This happens because `ProductOptionChoiceType::configureOptions` function tries to load all product options at once with `$this->productOptionRepository->findAll()`.
```
public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'choices' => function (Options $options): array {
                return $this->productOptionRepository->findAll();
            },
            'choice_value' => 'code',
            'choice_label' => 'name',
            'choice_translation_domain' => false,
        ]);
    }
```

### PR DESCRIPTION
**Pull request by ulinskas**
…ctOptionAutocompleteType

| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| Related tickets | fixes #13742                      |
| License         | MIT                                                          |

Before:
image

After:
image

### CODE REVIEW

**Comment by jakubtobiasz** [comment]
Hi @ulinskas!
This PR looks promising in terms of functionality. Would you like to fix Behat scenarios, as they're broken now? Meanwhile, I (or we) will analyze how much we break backward compatibility, as often removing any class is a bad sign.

**Comment by ulinskas** [comment]
@jakubtobiasz I will try to fix Behat scenarios

**Comment by TheMilek** [comment]
Hi @ulinskas! 🚀

Thank you for submitting the pull request 
I just wanted to confirm if you are still working on it and planning to complete the changes 🍻 

Please let us know if you have any questions or concerns 😄

**Comment by ulinskas** [comment]
@TheMilek 
Yes, I do. I'm currently on a long running vacations, but getting back to work in 2 weeks. So I hope, in a 4-5 weeks period, I will be able to introduce a solution for the tests.

**Comment by Rafikooo** [comment]
Hey @ulinskas,

This is a friendly reminder in case you forgot about it 😆 

Best wishes,
Rafał

File: src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductOptionRepository.php [hunk 1]
```diff
@@ -40,4 +41,40 @@ public function findByName(string $name, string $locale): array
             ->getResult()
         ;
     }
+
+    /**
+     * @return ProductOptionInterface[]
+     */
+    public function findByNamePart(string $phrase, string $locale, int $limit = 15): array
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductOptionRepository.php]:
```suggestion
    public function findByPhrase(string $phrase, string $locale, int $limit = 15): array
```
Fits better, we're also checking the code

File: src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductOptionRepository.php [hunk 2]
```diff
@@ -40,4 +41,40 @@ public function findByName(string $name, string $locale): array
...
+            ->getQuery()
+            ->getResult()
+        ;
+    }
+
+    /**
+     * @param string[]|null $code
+     * @return ProductOptionInterface[]
+     */
+    public function findByOptions(?array $code = []): array
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductOptionRepository.php]:
```suggestion
    public function findByCodes(array $codes = []): array
```
Don't see the reason for allowing null

File: src/Sylius/Component/Product/Repository/ProductOptionRepositoryInterface.php [hunk 3]
```diff
@@ -25,4 +25,15 @@ public function createListQueryBuilder(string $locale): QueryBuilder;
...
     public function findByName(string $name, string $locale): array;
+
+    /**
+     * @return ProductOptionInterface[]
+     */
+    public function findByNamePart(string $phrase, string $locale, int $limit = 15): array;
+
+    /**
+     * @param string[]|null $code
+     * @return ProductOptionInterface[]
```
Comment (NoResponseMate) [inline src/Sylius/Component/Product/Repository/ProductOptionRepositoryInterface.php]:
```suggestion
     * @param string[]|null $code
     *
     * @return ProductOptionInterface[]
```

File: src/Sylius/Bundle/ProductBundle/Form/Type/ProductType.php [hunk 4]
```diff
@@ -68,6 +68,11 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
             ->add('associations', ProductAssociationsType::class, [
                 'label' => false,
             ])
+            ->add('options', ProductOptionAutocompleteType::class, [
+                'required' => false,
+                'label' => 'sylius.form.product.options',
+                'multiple' => true,
+            ])
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ProductBundle/Form/Type/ProductType.php]: Most likely should be removed, it's added via the subscriber

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php [hunk 5]
```diff
@@ -46,7 +46,7 @@ public function preSetData(FormEvent $event): void
         /** Options should be disabled for configurable product if it has at least one defined variant */
         $disableOptions = (null !== $this->variantResolver->getVariant($product)) && $product->hasVariants();
 
-        $form->add('options', ProductOptionChoiceType::class, [
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php:49]: From what I see, the `ProductOptionChoiceType` is no longer used. As we can't remove it now (BC policy) let's deprecate it at least.

**Review by jakubtobiasz — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php (hunk 5 above)
Comment (NoResponseMate) [reply src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php:49]: The deprecation will go in a separate PR targeting 1.13

**Review by NoResponseMate — APPROVED** [review]

**Review by SzymonFilipek — APPROVED** [review]
Changes passed QA check

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @ulinskas!

**Pull request merged**